WIS-2660 | Add first-party subdomain support to gb-tracker-client cookie tracking

mattwelke opened the pull request:
Changes visitor ID cookie setting to use the apex domain when the hostname is detected to be a domain (aka not localhost). Requires testing by using on sites deployed to two domains.

Commit:
Changes visitor ID cookie setting to use the apex domain when the hostname is detected to be a domain (aka not localhost). Requires testing by using on sites deployed to two domains.

Commit:
Bump version to 4.0.1.

Commit:
Remove exports.

Commit:
typo

mattwelke commented on test/gb-tracker-core.spec.ts:
```diff
@@ -360,7 +360,7 @@ describe('gb-tracker-core tests', () => {
 
     const gbTrackerCore = new GbTracker('testcustomer', 'area');
-    const CookiesLib = gbTrackerCore.__getInternals().COOKIES_LIB;
+    const CookiesLib = gbTrackerCore.__getInternals().COOKIES_LIB as any;
```
The new `as any`s are a result of me making the type stronger in the main code. The legacy test code mutated the cookies library that was imported, which is no longer valid.

Commit:
Update version in README.md."

dependabot[bot] committed:
Bump @types/mocha from 7.0.2 to 9.0.0 (#100)

Bumps [@types/mocha](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/mocha) from 7.0.2 to 9.0.0.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/mocha)

---
updated-dependencies:
- dependency-name: "@types/mocha"
  dependency-type: direct:development
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Commit:
Remove package @groupby/beacon-models.

Commit:
Fix conflict.

PeterPandelidis-GBI commented on src/utils/index.ts:
```diff
@@ -47,3 +47,23 @@ export function startsWithOneOf(target: string, array: string[]): boolean {
+    }
+
+    const split = host.split('.');
+    return `${split[split.length - 2]}.${split[split.length - 1]}`;
```
How does this work for a co.uk domain

mattwelke replied:
Good call. Will add more test cases and fix the source code.

Commit:
Change tracker signature to use options object for override URL option instead of string param. Update major version.

Commit:
add docs

mattwelke commented:
Refactored the constructor so that the third param is an object instead of a string. Moved the existing logic for overriding the URL (which we use for testing) to this new options object. Updated docs too:

## Options

The constructor for the tracker client has a third, optional parameter for providing options:

```typescript
const tracker = new GbTracker('customer_id', 'area', {
    overrideUrl: '<some_url>' // Optional, overrides the URL the beacon is sent to. Useful for testing.
});
```

__________

Plan is to use this new parameter for all optional config moving forward, including the new cross-domain support. Functionality in this PR will do automatic cross-domain for all domains ending in `.com`. In the future, we'll add an option to this options param (as a non-breaking change) where people can put in the domains they use. This would override the automatic (for `.com` only) behavior for cross-domain.
